### PR TITLE
Delete crops

### PR DESCRIPTION
**Pull request by jamesgorrie**
We can now delete crops.

__Next__:
* Add action to media-api
* Expose in UI

TODO:
- [x] add permission
- [x] erroring
- [x] ~~throw s3 errors~~ TODO in another PR as it affect the whole `S3StorageFile`

### CODE REVIEW

**Comment by jamesgorrie** [comment]
Out of progress!

File: cropper/app/controllers/Application.scala [hunk 1]
```diff
@@ -107,6 +110,19 @@ object Application extends Controller with ArgoHelpers {
...
+    validateUserWithPermissions(httpRequest.user, PermissionType.DeleteCrops) map { user =>
+      Crops.deleteCrops(id).map { _ =>
+        Notifications.publish(Json.obj("id" -> id), "delete-image-exports")
+        Accepted
+      } recover {
+        case _ => respondError(BadRequest, "deletion-error", "Could not delete crops")
+      }
+    } recover {
+      case PermissionDeniedError => Future.successful(respondError(Unauthorized, "permission-denied", "You cannot delete crops"))
+    } flatMap (a => a) // <- flattening the futures for the response
```
Comment (theefer) [inline cropper/app/controllers/Application.scala]: Weird, we should not have to do that. Can we not use `flatMap` above?

File: cropper/app/lib/Permissions.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+import com.gu.mediaservice.lib.auth._
+
+
+object PermissionDeniedError extends Throwable("Permission denied")
+
+object Permissions {
+  val permissionStore = new PermissionStore(Config.configBucket, Config.awsCredentials)
+
+  def validateUserWithPermissions(user: Principal, permission: PermissionType.PermissionType)(implicit ec: ExecutionContext): Future[Principal] = {
```
Comment (theefer) [inline cropper/app/lib/Permissions.scala]: Wrap long line?

File: cropper/app/lib/Permissions.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    val failFuture = Future.failed(PermissionDeniedError)
```
Comment (theefer) [inline cropper/app/lib/Permissions.scala]: Could be hoisted outside this function to avoid creating it each time (even if successful)?

Comment (theefer) [inline cropper/app/lib/Permissions.scala]: Only works if the object is not stateful though.. I *think* that's safe here..?

File: cropper/app/lib/Permissions.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
+package lib
+
+import scala.concurrent.{ExecutionContext, Future}
+import com.gu.mediaservice.lib.auth._
+
+
+object PermissionDeniedError extends Throwable("Permission denied")
+
+object Permissions {
```
Comment (theefer) [inline cropper/app/lib/Permissions.scala]: Nice to have a helper for that!

**Comment by theefer** [comment]
A few comments but looks good!

File: cropper/app/controllers/Application.scala (hunk 1 above)
Comment (jamesgorrie) [inline cropper/app/controllers/Application.scala]: Yea, I was battling with the syntax a little. Updated.

File: cropper/app/lib/Permissions.scala (hunk 2 above)
Comment (jamesgorrie) [inline cropper/app/lib/Permissions.scala]: :+1:

File: cropper/app/lib/Permissions.scala (hunk 3 above)
Comment (jamesgorrie) [inline cropper/app/lib/Permissions.scala]: Safe.

File: cropper/app/lib/Permissions.scala (hunk 4 above)
Comment (jamesgorrie) [inline cropper/app/lib/Permissions.scala]: We can hopefully use this to do things like multiple permission etc later... probably move it over to common when / if we need to.

**Comment by theefer** [comment]
Nice :+1:

**Pull request merged**